core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,539 @@
+		reconnectTotalTimeThreshold:         getReConnectTotalTimeThreshold(),
+		reconnectBlockRcvTotalTimeThreshold: util.GetDurationOrDefault("peer.deliveryclient.bft.blockRcvTotalBackoffDelay", bftBlockRcvTotalBackoffDelay),
+		minBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.minBackoffDelay", bftMinBackoffDelay),
+		maxBackoffDelay:                     util.GetDurationOrDefault("peer.deliveryclient.bft.maxBackoffDelay", bftMaxBackoffDelay),
+		blockCensorshipTimeout:              util.GetDurationOrDefault("peer.deliveryclient.bft.blockCensorshipTimeout", bftBlockCensorshipTimeout),
+		endpoints:                           comm.Shuffle(endpoints),
+		blockReceiverIndex:                  -1,
+		headerReceivers:                     make(map[string]*bftHeaderReceiver),
+	}
+
+	bftLogger.Infof("[%s] Created BFT Delivery Client", chainID)
+	return c
+}
+
+func (c *bftDeliveryClient) Recv() (response *orderer.DeliverResponse, err error) {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	c.startOnce.Do(func() {
+		var num uint64
+		num, err = c.ledgerInfoProvider.LedgerHeight()
+		if err != nil {
+			return
+		}
+		c.nextBlockNumber = num
+		c.lastBlockTime = time.Now()
+		bftLogger.Debugf("[%s] Starting monitor routine; Initial ledger height: %d", c.chainID, num)
+		go c.monitor()
+	})
+	// can only happen once, after first invocation
+	if err != nil {
+		bftLogger.Debugf("[%s] Exit: error=%v", c.chainID, err)
+		return nil, errors.New("cannot access ledger height")
+	}
+
+	var numEP int
+	var numRetries int
+	var stopRetries = time.Now().Add(c.reconnectTotalTimeThreshold)
+
+	for !c.shouldStop() {
+		if numEP, err = c.assignReceivers(); err != nil {
+			bftLogger.Debugf("[%s] Exit: error=%s", c.chainID, err)
+			return nil, err
+		}
+
+		c.launchHeaderReceivers()
+
+		response, err = c.receiveBlock()
+		if err == nil {
+			bftLogger.Debugf("[%s] Exit: response=%v", c.chainID, response)
+			return response, nil // the normal return path
+		}
+
+		if stopRetries.Before(time.Now()) {
+			bftLogger.Debugf("[%s] Exit: reconnectTotalTimeThreshold: %s, expired; error: %s",
+				c.chainID, c.reconnectTotalTimeThreshold, errClientReconnectTimeout)
+			return nil, errClientReconnectTimeout
+		}
+
+		c.closeBlockReceiver()
+		numRetries++
+		if numRetries%numEP == 0 { //double the back-off delay on every round of attempts.
+			dur := backOffDuration(2.0, uint(numRetries/numEP), bftMinBackoffDelay, bftMaxBackoffDelay)
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint in: %s", c.chainID, err, dur)
+			backOffSleep(dur, c.stopChan)
+		} else {
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint now", c.chainID, err)
+		}
+	}
+
+	bftLogger.Debugf("[%s] Exit: %s", c.chainID, errClientClosing.Error())
+	return nil, errClientClosing
+}
+
+func backOffDuration(base float64, exponent uint, minDur, maxDur time.Duration) (backOffDur time.Duration) {
```

Comment:
isn't it possible to make this simpler?
You pass here a base, but the base is always 2.0 when you call the function, so why not just encode the base directly and not pass it as an argument? 
Also, if we have constants for max and minimum then why do you pass this in the function as parameters? 
If you just use the constants directly then you don't need to validate the input to the function as you do.

Now, I'm not sure I understand why the computation of the fDurNano is as it is. 
I think that if you want the time to be a multiple of the minimum delay then you can just divide the max delay by the min delay and then take this number to be the maximum of the exponentiation and then you don't need to multiply by nano seconds and have the corner cases of the infinity and what-not :)

Reply:
s/constant/field/g

Reply:
This is to avoid overflow when assigning a floating point to a int64.